### PR TITLE
IPFS Integrity Check

### LINKED ISSUE #177: IPFS Upload Integrity Check — add + cat
needed-by #53 
needed-by #112 

# Tasks

- [x] RDD
- [x] There are ~20% broken files in production. Find a way to reproduce the bug locally, achieving similar results. We'll need a script that we can run to massively post files to the node.
- [x] Ideally, find a way to reproduce the bug for a single claim deterministically.
- [x] Break IPFS Upload out of Storage module, separating Read and Write operations into two modules: `StorageReader` and `StorageWriter`.
- [x] Make IFPS Upload async, similar to our current IPFS Download process.
- [ ] Add integrity check to `StorageWriter`, storing the result of the check in the database.

### PR DESCRIPTION
**Pull request by kennylavender**
[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

closes #177 

Add integrity check step to the claim controller. The point is so that we can be notified if a claim was not correctly uploaded to IPFS and record the information so we can have a better reason as to why the claim was not correctly uploaded IPFS.

### Make it possible for claims to break
Adjust the IPFS helper so it will break with non-ascii characters. Change this line

https://github.com/poetapp/node/blob/552b4b81acb778f9cf32d8723e7922e381e0b978/src/Helpers/IPFS.ts#L40
to
```
knownLength: text.length,
```
### Adjust your docker-compose.override.yml file
```yml
version: '3.6'

services:
  poet-node:
    volumes:
      - .:/usr/src/app
    command: sleep infinity
```

#### Run the node and requirements
* term1: `docker-compose up -d`
* term1: `docker-compose exec poet-node bash`
* term1: `npm run build && npm run start`

#### Post a working claim

* term2 `npm run create-claim` then wait for the log that the claim was stored succesfully

#### Post a claim that will trigger the Integrity check to fail
Create a claim with non-ascii characters that will fail due to our purpose built failure in the IPFS helper. You should see that a IntegrityCheckFailure was thrown. The name we pass create-claim needs to be long enough to cause a length mismatch for the stream

* term2 `npm run create-claim 火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了Индейка нашла свой путь домойИндейка нашла свой путь домойИндейка нашла свой путь домойИндейка нашла свой путь домойИндейка нашла свой путь домойИндейка нашла свой путь домойИндейка нашла свой путь домойИндейка нашла свой путь домойИндейка нашла свой путь домой`

#### View the results
* term3: `docker-compose exec mongo bash`
* term3: `mongo --username poet --password poetPass poet`
* term3: `db. db.storageWriterIntegrityCheckFailures.find().pretty()` // Should give you back the integrity check failure details.

#### Example failure from the database
```
{
	"_id" : ObjectId("5be4d289ee5bc101be5f905a"),
	"claim" : {
		"@context" : {
			"cred" : "https://w3id.org/credentials#",
			"dc" : "http://purl.org/dc/terms/",
			"schema" : "http://schema.org/",
			"sec" : "https://w3id.org/security#",
			"id" : "sec:digestValue",
			"issuer" : "cred:issuer",
			"issuanceDate" : "cred:issued",
			"type" : "schema:additionalType",
			"claim" : "schema:CreativeWork",
			"archiveUrl" : "schema:url",
			"author" : "schema:author",
			"canonicalUrl" : "schema:url",
			"contributors" : {
				"@id" : "schema:ItemList",
				"@container" : "@list",
				"@type" : "schema:contributor"
			},
			"copyrightHolder" : "schema:copyrightHolder",
			"dateCreated" : "schema:dateCreated",
			"datePublished" : "schema:datePublished",
			"license" : "schema:license",
			"name" : "schema:name",
			"tags" : "schema:keywords",
			"hash" : "sec:digestValue"
		},
		"id" : "3d176f0f8c65a73f3d380e52b7641a9c81f700058599f6662544de87b584b04e",
		"type" : "Work",
		"issuer" : "data:;base64,eyJhbGdvcml0aG0iOiJFZDI1NTE5U2lnbmF0dXJlMjAxOCIsInB1YmxpY0tleSI6IkdhRWZ2QURLQUw1ZXVWQjZxZ2p1djlnMkxoOVBhM2FuWkxLZjRnUlFvWVM0In0=",
		"issuanceDate" : "2018-11-09T00:17:20.099Z",
		"claim" : {
			"name" : "火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了Индейка",
			"author" : "нашла"
		},
		"sec:proof" : {
			"@graph" : {
				"@type" : "sec:Ed25519Signature2018",
				"dc:created" : {
					"@type" : "http://www.w3.org/2001/XMLSchema#dateTime",
					"@value" : "2018-11-09T00:17:20Z"
				},
				"dc:creator" : {
					"@id" : "data:;base64,eyJhbGdvcml0aG0iOiJFZDI1NTE5U2lnbmF0dXJlMjAxOCIsInB1YmxpY0tleSI6IkdhRWZ2QURLQUw1ZXVWQjZxZ2p1djlnMkxoOVBhM2FuWkxLZjRnUlFvWVM0In0="
				},
				"sec:jws" : "eyJhbGciOiJFZERTQSIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..w-aec-5-Re_QDaY7iO258d62P_9q5ngmwBOaVouKdBf85eftD8AiuGpbJdL5qbhcpeSC9ws0DWICxn0TuufXAg",
				"sec:nonce" : "cjo99uqjx0000fiy62qy1kr9h"
			}
		}
	},
	"ipfsResponse" : "{\"@context\":{\"cred\":\"https://w3id.org/credentials#\",\"dc\":\"http://purl.org/dc/terms/\",\"schema\":\"http://schema.org/\",\"sec\":\"https://w3id.org/security#\",\"id\":\"sec:digestValue\",\"issuer\":\"cred:issuer\",\"issuanceDate\":\"cred:issued\",\"type\":\"schema:additionalType\",\"claim\":\"schema:CreativeWork\",\"archiveUrl\":\"schema:url\",\"author\":\"schema:author\",\"canonicalUrl\":\"schema:url\",\"contributors\":{\"@id\":\"schema:ItemList\",\"@container\":\"@list\",\"@type\":\"schema:contributor\"},\"copyrightHolder\":\"schema:copyrightHolder\",\"dateCreated\":\"schema:dateCreated\",\"datePublished\":\"schema:datePublished\",\"license\":\"schema:license\",\"name\":\"schema:name\",\"tags\":\"schema:keywords\",\"hash\":\"sec:digestValue\"},\"id\":\"3d176f0f8c65a73f3d380e52b7641a9c81f700058599f6662544de87b584b04e\",\"type\":\"Work\",\"issuer\":\"data:;base64,eyJhbGdvcml0aG0iOiJFZDI1NTE5U2lnbmF0dXJlMjAxOCIsInB1YmxpY0tleSI6IkdhRWZ2QURLQUw1ZXVWQjZxZ2p1djlnMkxoOVBhM2FuWkxLZjRnUlFvWVM0In0=\",\"issuanceDate\":\"2018-11-09T00:17:20.099Z\",\"claim\":{\"name\":\"火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了火雞已回家了Индейка\",\"author\":\"нашла\"},\"sec:proof\":{\"@graph\":{\"@type\":\"sec:Ed25519Signature2018\",\"dc:created\":{\"@type\":\"http://www.w3.org/2001/XMLSchema#dateTime\",\"@value\":\"2018-11-09T00:17:20Z\"},\"dc:creator\":{\"@id\":\"data:;base64,eyJhbGdvcml0aG0iOiJFZDI1NTE5U2lnbmF0dXJlMjAxOCIsInB1YmxpY0tleSI6IkdhRWZ2QURLQUw1ZXVWQjZxZ2p1djlnMkxoOVBhM2FuWkxL",
	"message" : "Unexpected end of JSON input",
	"failureTime" : ISODate("2018-11-09T00:19:21.121Z")
}
```

### CODE REVIEW

File: src/StorageWriter/IPFS.ts [hunk 1]
```diff
@@ -37,4 +38,11 @@ export class IPFS {
 
     return json.Hash
   }
+
+  cat = async (hash: string): Promise<string> => {
```
Comment (lautarodragan) [inline src/StorageWriter/IPFS.ts]: Let's avoid duplicating this one and refactor the one used by StorageReader.

File: src/StorageWriter/DAOClaims.ts [hunk 2]
```diff
@@ -44,6 +44,10 @@ export class DAOClaims {
     await this.collection.updateOne({ 'claim.id': claimId }, { $set: { ipfsFileHash } })
   }
 
+  public readonly getClaimByHash = (ipfsFileHash: string) => {
```
Comment (lautarodragan) [inline src/StorageWriter/DAOClaims.ts]: I think we're using `find` instead of `get` for DB operations.

**Comment by kennylavender** [comment]
blocked by #705

**Comment by kennylavender** [comment]
#705 was closed, now blocked by #717

**Comment by kennylavender** [comment]
@lautarodragan ready for review again.

**Comment by krobi64** [comment]
Why does the manual test involve modifying the code and then submitting a claim? What is this supposed to protect against?

**Comment by kennylavender** [comment]
@krobi64 we had issues with uploads to ipfs being cut off #53, the cause ended up being an incorrect length of the stream for non-ascii characters, which is why I used it as a way to test this change manually. This change just alerts us to any future issues we may have and records the information so we can debug the issue better. It was painfully time-consuming debugging #53

File: src/StorageWriter/DAOIntegrityCheckFailures.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import { Collection, Db} from 'mongodb'
+
+@injectable()
+export class DAOIntegrityCheckFailures {
+  private readonly collection: Collection
+
+  constructor(
+    @inject('DB') db: Db,
+  ) {
+    this.collection = db.collection('storageWriterIntegrityCheckFailures')
```
Comment (lautarodragan) [inline src/StorageWriter/DAOIntegrityCheckFailures.ts]: Let's inject the collection directly and move `db.collection('storageWriterIntegrityCheckFailures')` to the DI container creation

**Comment by kennylavender** [comment]
@lautarodragan @krobi64 If I move this logic to another file `ClaimIntegrityChecker.ts` or something we shouldn't have an issue mocking IPFS. I started doing this originally but wasn't completely sure the logic should be separated from the `ClaimController`, it felt a bit weird but it's probably the correct better thing to do.

**Pull request merged**